### PR TITLE
Python 3 support

### PR DESCRIPTION
**Pull request by lukebakken**
Fixes elastic/beats#3166

Test environment:

Arch Linux
Python `3.6.1` installed via `pyenv`
```
$ docker --version
Docker version 17.05.0-ce, build 89658bed64
```

Tested by running `make check && make`

### CODE REVIEW

**Comment by ruflin** [comment]
@lukebakken Thanks for taking this on. We should make sure to also keep it compatible with 2.7 at the same time.

**Comment by lukebakken** [comment]
@ruflin of course. Based on the Travis CI output it looks like the worker has a pretty old Python version. I'll make sure to be compatible with that. Thankfully AppVeyor is using the latest `2.7`.

**Comment by ruflin** [comment]
Based on our CONTRIBUTING.md it should be at least 2.7.9: https://github.com/elastic/beats/blob/master/CONTRIBUTING.md Are some relevant changes happening between 2.7.9 and the most recent 2.7 version? We could also up the requirement there if needed potentially to 2.7.x

**Comment by lukebakken** [comment]
I don't think there are relevant changes. The Travis CI output doesn't show what Python version is being used, I just know that `InsecurePlatformWarning` means "ancient Python". I'll add something so the version is echoed on the next build and will go with that.

File: dev-tools/generate_notice.py [hunk 1]
```diff
@@ -10,7 +11,7 @@ def read_file(filename):
         print("File not found {}".format(filename))
         return ""
 
-    with open(filename, 'r') as f:
+    with open(filename, 'rb') as f:
```
Comment (lukebakken) [inline dev-tools/generate_notice.py]: NB: one of the license files generated this error on read using Python 3:

```
UnicodeDecodeError: 'utf-8' codec can't decode byte 0x93 at position ...
```

Switching file operations to binary mode fixes this issue, but all strings must be bytes, necessitating the `six.b` additions.

**Comment by lukebakken** [comment]
If you're curious, what prompted all this was me running `make` on a fresh clone of your repository on my Arch Linux workstation. Python 3 is the system version OOTB. I also use `pyenv` so it was easy to switch to using Python 2 to run the test suite.

File: filebeat/tests/system/config/filebeat.yml.j2 [hunk 2]
```diff
@@ -32,10 +32,10 @@ filebeat.prospectors:
   {%- if prospector_processors %}
   processors:
     {%- for processor in prospector_processors %}
-    {%- for name, settings in processor.iteritems() %}
+    {%- for name, settings in processor.items() %}
```
Comment (lukebakken) [inline filebeat/tests/system/config/filebeat.yml.j2]: Both Python 2 and 3 have `.items()`. If adding an `import` within `{% %}` is acceptable, `six.iteritems` could be used.

**Comment by lukebakken** [comment]
OK tests pass locally using `3.6.1` save for one `protocol not supported` error returned from docker for [this test](https://github.com/elastic/beats/blob/master/metricbeat/module/audit/kernel/audit_linux_test.go#L13-L33). I'm assuming it's related to my docker version as the test passes on Travis CI.

**Comment by ruflin** [comment]
I think it would be nice if we could all go to python 3.x but there are still too many systems out there that have 2.7 as default :-(

Your changes seem to break our windows builds as the package `six` is missing.

File: filebeat/tests/system/config/filebeat.yml.j2 (hunk 2 above)
Comment (ruflin) [inline filebeat/tests/system/config/filebeat.yml.j2]: I think that should be ok.

**Comment by ruflin** [comment]
BTW: For the windows tests your have to check the appveyor file.

**Comment by lukebakken** [comment]
Yep, I'll get it. Yesterday I logged off prior to the AppVeyor build finishing.

**Comment by lukebakken** [comment]
Looks like everything's OK except the Jenkins Windows builds. It just needs to have `six` added in the same manner [as here](https://github.com/lukebakken/beats/blob/fixes/lrb/gh-3166-python-3-dict/.appveyor.yml#L58)

**Comment by ruflin** [comment]
@andrewkroh Can you help here on where this would have to be added?

File: Makefile [hunk 3]
```diff
@@ -1,11 +1,11 @@
 
 BUILD_DIR=build
-COVERAGE_DIR=${BUILD_DIR}/coverage
+COVERAGE_DIR=$(BUILD_DIR)/coverage
```
Comment (ruflin) [inline Makefile]: Is there a specific reason for changing `{´ to `(`. Nothing agains it more curious as it does not really belong to this PR.

File: filebeat/scripts/docs_collector.py [hunk 4]
```diff
@@ -1,6 +1,7 @@
 import os
 import argparse
 import yaml
+import six
```
Comment (ruflin) [inline filebeat/scripts/docs_collector.py]: why is six also needed here? as this does not seem to be related to the license files

File: Makefile (hunk 3 above)
Comment (lukebakken) [inline Makefile]: I saw that about half used `{ }` and half used `( )`, and did a quick `:%s/\${\([^}]\+\)}/$(\1)/g` on it. I'll revert it to just the necessary change.

File: filebeat/scripts/docs_collector.py (hunk 4 above)
Comment (lukebakken) [inline filebeat/scripts/docs_collector.py]: There are two instances of `six.iteritems` in that file ([here](https://github.com/lukebakken/beats/blob/fixes/lrb/gh-3166-python-3-dict/filebeat/scripts/docs_collector.py#L63) and [here](https://github.com/lukebakken/beats/blob/fixes/lrb/gh-3166-python-3-dict/filebeat/scripts/docs_collector.py#L68))

File: Makefile (hunk 3 above)
Comment (ruflin) [inline Makefile]: I'm very happy to have it unified, but lets do it in a different PR. Could you open an other one for that?

File: filebeat/scripts/docs_collector.py (hunk 4 above)
Comment (ruflin) [inline filebeat/scripts/docs_collector.py]: Ok, I initially thought based on your comment above that you need six for the encoding. But I read up a bit more on it and it is for the 2x to 3x compatibilitly. Got it.

File: Makefile (hunk 3 above)
Comment (lukebakken) [inline Makefile]: Certainly! I should have done that in the first place.

File: filebeat/scripts/docs_collector.py (hunk 4 above)
Comment (lukebakken) [inline filebeat/scripts/docs_collector.py]: Yeah, it's cleverly named `2 * 3 = six` 😏

**Comment by andrewkroh** [comment]
> @andrewkroh Can you help here on where this would have to be added?

Working on it. Long term I need to start using virtualenv on Windows for this.

**Comment by ruflin** [comment]
@lukebakken There seem to be some conflicts now. Could you rebase on master and push a new version?

**Comment by ruflin** [comment]
@lukebakken Thanks a lot for this contribution. Now as a next step comes the tricky part: How do we ensure that we keep it compatible with python3 :-)

File: dev-tools/generate_notice.py (hunk 1 above)
Comment (tsg) [inline dev-tools/generate_notice.py]: Just FYI, I went for a different solution in https://github.com/elastic/beats/pull/4383, to avoid all the `six.b` calls.

Comment (lukebakken) [inline dev-tools/generate_notice.py]: I see that. It's much better than a bunch of compatibility code.
